Single service mode

lexoyo opened the pull request:
closes #62

it is a simplified version of #62

clemos commented on src/js/CloudExplorer.jsx:
```diff
@@ -162,6 +162,15 @@ export default class CloudExplorer extends React.Component {
           });
+          // the first display, single service mode, try to enter
+          // this is useful when CE is used by hosting companies to display the user files, and the user has logged in their system
+          if(!this.initDone && files.length === 1 && path.length === 0/* && files[0].isLoggedIn === true*/) {
```
I think initDone should be part of the component state.

clemos replied:
but more generally, I don't quite understand this part.

lexoyo replied:
yes again that's a way to open a service when CE starts, so only the 1st time and if it is "/" which is `ls`
maybe i should put all this somewhere else...

lexoyo commented on src/js/UnifileService.js:
```diff
@@ -299,6 +299,7 @@ export default class UnifileService {
   static isService (file) {
-    return typeof file.isLoggedIn !== 'undefined';
+    console.log('isService', file, file.isLoggedIn);
+    return typeof file.mime === 'application/json';
```
the console log is a mistake

do you think this is ok for a comment? 

> this mime type is what the server returns when listing services... and I think it never will be the mime type of any other files

Do you think I should rather add a specific field "isService" to the data returned by the server side?

JbIPS replied:
I see a lot of use case when a regular file will have a `application/json` MIME type, so yes, it should already be in the data without duck-typing

lexoyo replied:
Ok, so do you think I should re-open this PR?
https://github.com/silexlabs/CloudExplorer2/pull/50

lexoyo replied:
> I see a lot of use case when a regular file will have a application/json MIME type, so yes, it should already be in the data without duck-typing

for a static file this would be strange
if you remember we decided that this would be the mime type of services, which make them apear with an icon of service

JbIPS replied:
You're thinking too much with Silex, not CE. As a developer, my FTP might be full of JSON config files I saved here.

You wouldn't want that to mess up CE

lexoyo commented on src/js/UnifileService.js:
```diff
@@ -124,9 +124,9 @@ export default class UnifileService {
   }
 
-  cd (path) {
+  cd (path, preventAuth=false) {
```
in fact  some services won't require auth, but this is about another topic here: when CE starts, if the path is "/" it will list services, and if there is only one service we want to `cd` into this folder. But without auth because we don't want a popup to be opened and blocked (not in a onclick). What happen here is we `cd` the service and if the user is not logged in it will return a `403 forbidden` and we will be back to "/"

The best thing would be to send a `isLoggedIn` field with the services, so if the user is loggedin we `cd` in it, and when we `cd` a service to which we are logged in, CE should not bother opening a popup for auth. I made a PR for that but I think it died forgotten.. Maybe it was wrong :-/

lexoyo commented on src/js/UnifileService.js:
```diff
@@ -124,9 +124,9 @@ export default class UnifileService {
+  cd (path, preventAuth=false) {
     return new Promise((resolve, reject) => {
-      if (path.length === 1 && path[0] !== this.currentPath[0]) {
+      if (!preventAuth && path.length === 1 && path[0] !== this.currentPath[0]) {
```
yeah.. it's a way to check if we are opening a service or a folder... we should probably use isService or something
but we don't have the files yet at this level, only the path
any suggestions?

JbIPS replied:
We could retrieve the service with the path, right? If it's the first part of the path, use the service `Map` to get the service and fetch the service auth strategy (I think that's what @clemos was trying to say)

lexoyo replied:
ok, then that's what `path.length === 1` does

lexoyo commented:
Ok so
* I have fixed the issues you raised
* I have added a "isLoggedIn" info and "isService" info as I previously wanted to do in the old PR
* since this prevented any call to Dropbox or Github when the user had been logged in once, I had to add a logout button - which I did at the service level in the UI, but a "logout all" button somewhere might be needed by @godzone for the single service use case

lexoyo commented:
![ce2-logout-single-service](https://user-images.githubusercontent.com/715377/44626648-5c74b480-a920-11e8-98ac-233d4989814f.gif)

clemos commented on src/js/CloudExplorer.jsx:
```diff
@@ -162,6 +162,18 @@ export default class CloudExplorer extends React.Component {
           });
+          // the first display, single service mode, try to enter
+          // this is useful when CE is used by hosting companies to display the user files, and the user has logged in their system
+          if(!this.initDone && files.length === 1 &&
```
I've had a deeper look, and here's what I suggest:
* implement a `getServices` in `UnifileService`
* move initialization of services out of this component, maybe in App.jsx
* `initDone` should be part of a state, somewhere, and you should pass a `services` map as props to a component (maybe `CloudExplorer`)

what do you think ?

lexoyo replied:
Yes of course! Good idea

Just to be sure, what do you mean by initialization of services" ?
You mean App.js should start by requesting a list of services and then instanciate the components with this list?

I need to figure out how app.js will be notified when the user logs in successfully to a service, in in order to update the services in props

clemos replied:
> Just to be sure, what do you mean by initialization of services" ?
You mean App.js should start by requesting a list of services and then instanciate the components with this list?
>
> I need to figure out how app.js will be notified when the user logs in successfully to a service, in in order to update the services in props

Yes, it would be something like that : 
* app initializes and loads the list of services (view is just a loading spinner)
* top level component gets the list of services via props (you can consider the list of services a configuration, except it's loaded at initialization)

This way, you can break the logic where you `cd` to do anything:
Outer component is responsible for:
* listing services
* authenticating / logout if necessary
Inner component is responsible for 1 service.

Outer component state: 
```
{ 
   loading: boolean, // indicates the list of services is loading
   authenticating: boolean, // or { [serviceId]: boolean } maybe
   services: {
      // id to service map, more or less as returned by the server
   },
   currentService: // serviceId or null,
   auth : {
     // can be a simple <serviceId> to boolean map
     // it is not in each services object to avoid merging objects in services / mixing "data" and state
   },
   error: // null or error to display
}
```
allows to : 
```
// --
this.setState({loading: true, auth: {}}) // probably clear auth map (?)
loadServices(url)
   .then(res => this.setState({services: res.services, loading: false})
   .catch(e => this.setState({error:e, loading: false})
// --
auth(serviceId)
   .then(res => this.setState({auth: {[serviceId]: true}}));
// --
```

Well I don't know if it makes things clearer to you, but
&tldr; receiving services, receiving successful login MUST trigger a `setState()` (resulting in a rerender)

Then the `state` of the "inner" component could be like:
```js
{
   files: [], // list of files currently displayed
   currentPath : '/toto', // ...
   loading: boolean
}
```
anything else is passed as props:
```
<FileService 
    serviceId={serviceId} // you can clear inner state when detecting changes on that
    service={service} // promise based API, like service.ls(path).then(files=>this.setState({files, loading: false})
    onLoggedOut={onLoggedOut} // just in case of auth timeout, but this is another story
/> 
```

Finally, you have a stateless "item list" component that manages displaying files and services in a similar way.

That's probably not perfect, but I hope it helps.

clemos replied:
And then you will reply:
How to effectively display a breadcrumb where service and path are mixed :D ?
That's where proper / global state management arrives in the discussion, and I advise you to give Redux a try :D

clemos replied:
Final word:

Don't ever.
Ever.
Ever.
Store state in a Component member variable like `this.initDone`.
99.999999% of the time, you want to trigger a re-render when doing this, and thus you want to do `this.setState({initDone:true})`

lexoyo replied:
ok thx, this is a lot to refactor where I just needed a quick fix, but I take the advice and will give it a try

for redux, maybe you could start integrating it yourself? I need to maintain Silex, I can't spend too much time on CE sadly

lexoyo replied:
so, as I told you @clemos I only moved my flag `initDone` to the component's state
a refactoring as you suggested seemed good to me at first, but the logic inside UnifileService is completely changed and at the end I have as much code and it seemed more complicated to me

also @JbIPS I tried to fix some linter errors, but almost every file has warning and errors... I had to force commits with `-n`, can you look at this and tell me what happened since you told me you had no errors when you installed the linter... maybe a config problem, or your local config different from the one in the repo?

anyway I merge this, I have test more than once, because unit testing is not enough for pros like us ;)

clemos replied:
I've started refactoring the app here : 
https://github.com/silexlabs/CloudExplorer2/compare/refacto-react?expand=1#diff-9f05fb0c0e4d64f24f28eec0fce0d8f4
I don't know if I will finish it ever, but at least I hope it explains better how you are supposed to deal with state in React

clemos replied:
Oh, and also: fuck pre-commit hooks !!!!!

lexoyo commented:
oh there is an old issue about the linter: #45

lexoyo commented:
+1

Le 1 sept. 2018 à 14:10, à 14:10, "Clément Charmet" <notifications@github.com> a écrit:
>clemos commented on this pull request.
>
>
>
>> @@ -162,6 +162,18 @@ export default class CloudExplorer extends
>React.Component {
>             files,
>             loading: false
>           });
>+          // the first display, single service mode, try to enter
>+          // this is useful when CE is used by hosting companies to
>display the user files, and the user has logged in their system
>+          if(!this.initDone && files.length === 1 &&
>
>Oh, and also: fuck pre-commit hooks !!!!!
>
>-- 
>You are receiving this because you modified the open/close state.
>Reply to this email directly or view it on GitHub:
>https://github.com/silexlabs/CloudExplorer2/pull/68#discussion_r214512269
